KAFKA-7961: Ignore assignment for un-subscribed partitions

jsancio opened the pull request:
Whenever the consumer coordinator sends a response that doesn't match the client consumer subscription, ignore the assignment and rejoin the group.

Testing strategy: create a mocked client that first sends an assignment response that doesn't match the client subscription followed by an assignment response that does match the client subscription.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jsancio committed:
KAFKA-7961: Ignore assignment for un-subscribed partitions

jsancio committed:
KAFKA-7961: Test rejoin with incorrect assignment

Added "testInvalidCoordinatorAssignment" test wich result in the
following failure:

```
org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest > testInvalidCoordinatorAssignment FAILED
    java.lang.IllegalArgumentException: Assigned partition test2-0 for non-subscribed topic; subscription is [test1]
        at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignFromSubscribed(SubscriptionState.java:195)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:249)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:410)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:344)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:340)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest.testInvalidCoordinatorAssignment(ConsumerCoordinatorTest.java:462)

1 test completed, 1 failed
```

jsancio commented:
cc @guozhangwang @ijuma @hachikuji

jsancio commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -414,6 +414,61 @@ public boolean matches(AbstractRequest body) {
     }
 
+    @Test
+    public void testInvalidCoordinatorAssignment() {
```
Wrote the test first as it failed with this error just as described in the issue:
```
org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest > testInvalidCoordinatorAssignment FAILED
    java.lang.IllegalArgumentException: Assigned partition test2-0 for non-subscribed topic; subscription is [test1]
        at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignFromSubscribed(SubscriptionState.java:195)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:249)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:410)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:344)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:340)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest.testInvalidCoordinatorAssignment(ConsumerCoordinatorTest.java:462)

1 test completed, 1 failed
```

hachikuji replied:
I think there are two possible scenarios for this case:

1. The coordinator has assigned partitions that we did not ask for.
2. The subscription has changed while the JoinGroup was in-flight.

When I filed the bug, it was mostly the second scenario that I was thinking about. If the subscription has changed, then it is reasonable to request another rebalance. However, in the first case, raising the exception may have been a better approach since it likely means a bug in the assignment logic. If we do not do so, we may get stuck in a rebalance loop.

In fact, we keep track of the subscription that was sent in the JoinGroup through `ConsumerCoordinator.joinedSubscription`. Do you think it would make sense to validate the assigned partitions against that set in the case that `assignFromSubscribed` returns false? If there are assigned partitions that we didn't ask for, we can still raise an exception.

jsancio replied:
Done.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -174,28 +177,45 @@ public void assignFromUser(Set<TopicPartition> partitions) {
-                    throw new IllegalArgumentException("Assigned partition " + tp + " for non-subscribed topic; subscription is " + this.subscription);
+        };
+
+        boolean assignmentMatchedSubscrition = assignments.stream().allMatch(predicate);
```
typo: "Subscrition"

jsancio replied:
Done.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -56,6 +58,7 @@
 public class SubscriptionState {
     private static final String SUBSCRIPTION_EXCEPTION_MESSAGE =
             "Subscription to topics, partitions and pattern are mutually exclusive";
+    private static final Logger log = LoggerFactory.getLogger(SubscriptionState.class);
```
Can we pass `LogContext` from the consumer to construct this? Often applications have many consumers and it becomes hard to distinguish individual consumer behavior, so we created a log context which basically just adds a prefix to all log messages.

jsancio replied:
Done.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -174,28 +177,45 @@ public void assignFromUser(Set<TopicPartition> partitions) {
+     * @return true if assignments matches subscription, otherwise false
      */
-    public void assignFromSubscribed(Collection<TopicPartition> assignments) {
+    public boolean assignFromSubscribed(Collection<TopicPartition> assignments) {
```
It may be helpful to have some focused unit tests in `SubscriptionStateTest`.

jsancio replied:
Good catch. We have many tests in `SubscriptionStateTest`. I updated all of them to assert the correct value is returned.

hachikuji reviewed (commented):
Thanks for picking this up. Looks good overall. Left a few comments.

jsancio committed:
KAFKA-7961: Add LogContext to SubscriptionState

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -174,28 +179,45 @@ public void assignFromUser(Set<TopicPartition> partitions) {
+            if (this.subscribedPattern != null) {
+                boolean match = this.subscribedPattern.matcher(topicPartition.topic()).matches();
+                if (!match) {
+                    log.info("Assigned partition " + topicPartition +
```
nit: use `{}` substitutions? Same below.

jsancio replied:
Done.

jsancio committed:
KAFKA-7961: Handle invalid response from leader

Throw an IllegalStateException if the leader sent and assignment that
doesn't match for the subscription requested.

Modify testInvalidCoordinatorAssignment and add
testOutdatedCoordinatorAssignmetn tests for cover this condition.

jsancio commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -414,6 +415,98 @@ public boolean matches(AbstractRequest body) {
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void testInvalidCoordinatorAssignment() {
```
For this test, the old code fails as follow:
```
org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest > testInvalidCoordinatorAssignment FAILED
    java.lang.Exception: Unexpected exception, expected<java.lang.IllegalStateException> but was<java.lang.IllegalArgumentException>

        Caused by:
        java.lang.IllegalArgumentException: Assigned partition test2-0 for non-subscribed topic; subscription is [test1]
            at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignFromSubscribed(SubscriptionState.java:195)
            at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:249)
            at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:410)
            at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:344)
            at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:340)
            at org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest.testInvalidCoordinatorAssignment(ConsumerCoordinatorTest.java:506)
```

jsancio commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -414,6 +415,98 @@ public boolean matches(AbstractRequest body) {
     }
 
+    @Test
+    public void testOutdatedCoordinatorAssignment() {
```
For this test the old code fails as follow:
```
org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest > testOutdatedCoordinatorAssignment FAILED
    java.lang.IllegalArgumentException: Assigned partition test2-0 for non-subscribed topic; subscription is [test1]
        at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignFromSubscribed(SubscriptionState.java:195)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:249)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:410)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:344)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:340)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest.testOutdatedCoordinatorAssignment(ConsumerCoordinatorTest.java:466)
```

jsancio reviewed (commented):
@hachikuji Here are the failures when we run these tests against the old code.

jsancio committed:
KAFKA-7961: Fix logging to use "{}"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -246,7 +247,17 @@ protected void onJoinComplete(int generation,
+            Set<TopicPartition> invalidAssignments = assignment.partitions().stream().filter(topicPartition -> 
+                !joinedSubscription.contains(topicPartition.topic())).collect(Collectors.toSet());
+            if (invalidAssignments.size() > 0) {
+                throw new IllegalStateException("Coordinator leader sent assignment that don't correspond to subscription request: " + invalidAssignments.toString());
```
nit: `toString` can be used implicitly

jsancio replied:
Done.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -414,6 +415,98 @@ public boolean matches(AbstractRequest body) {
+                        sync.groupAssignment().containsKey(consumerId);
+            }
+        }, syncGroupResponse(singletonList(t2p), Errors.NONE));
+        coordinator.poll(time.timer(Long.MAX_VALUE));
```
I found out recently about the fancy pants `assertThrows` api. It seems like a nicer pattern because you can localize the block of code that should be raising the exception. 

As an aside, are there any assertions on the state of the coordinator falling raising this exception? Maybe at least we could verify the assignment in `SubscriptionState` is empty. I think generally this should be viewed by users as a fatal error.

jsancio replied:
Changing it to use `assertThrows`.

Regarding checking additional state, I am hesitant to check for certain state after we throw an `IllegalStateException`. It is my opinion that if the test checks for certain values we are stating to the user that this is a valid state after `poll` throws such an exception. I think in this case the protocol was violated and the consumers/coordinators need to be re-created, right?

hachikuji replied:
Fair enough. So basically if we have reached an illegal state, the user continues at their own peril. :wink:

hachikuji reviewed (commented):
Thanks, the fix looks good. Just had a couple small comments.

jsancio committed:
KAFKA-7961: Use assertThrows

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

ijuma commented:
Should this be cherry-picked to 2.2?

jsancio commented:
@ijuma I don't think so. This was found during internal testing by waking up the consumer and while also changing the subscription.

@hachikuji What do you think?
